README.txt
----------------------------------------------------------------------------------------------------
eXtplorer 2.1.0 - README
----------------------------------------------------------------------------------------------------

Requirements:
-------------------
* PHP > 4.3

Supported Browsers:
-------------------
* Internet Explorer >= 6.0
* Firefox >= 2
* Safari >= 2
* Google Chrome/Iron >= 1
* Konqueror (KDE > 3.5)
* Opera 9+

Opera and Konqueror don't allow custom Context Menus. Use Double Click on a record in the grid instead.

Installation as Standalone Script
-------------------
	1. Unpack the Archive (where this README.txt is contained) to a local directory
	2. Login to your server via FTP and create a subdirectory for eXtplorer
	3. Upload all eXtplorer files from your computer into the subdirectory on the server
	4. Browse to the URL http://<YOURSERVER>/<SUBDIRECTORY> and login. Remember to immediately change the admin password.
	5. If you can't change your admin password, this is most likely because the user file is not writable. With your FTP program browse to
		the eXtplorer directory /config and make the file ".htusers.php" writable (CHMOD 666).
	
  Done.
  
Installation into Joomla!/Mambo:
-------------------
	1. Login to Joomla!'s Administration Backend /administrator
	2. Go to "Components", "Install/Uninstall".
	3. Browse to the file "com_eXtplorer_xx.tar.gz" on your computer and click on "Upload File & Install".
	
  Done.
  You can now access eXtplorer through "Components" => "eXtplorer"


License:
--------------------
The software is dual-licensed and subject to the 
	- Mozilla Public License Version 1.1 (the "License") 
	or 
	- to the terms of the GNU General Public License Version 2 or later (the "GPL"); 
	
You may not use this software except in compliance with one of these Licenses. 

You may obtain a copy of the Mozilla Public License at http://www.mozilla.org/MPL/.
You may obtain a copy of the GNU General Public License at http://www.gnu.org/copyleft/gpl.html.

Software distributed under the License is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
License for the specific language governing rights and limitations under the License.

Alternatively, the software may be used under the terms of the GNU General Public License Version 2 or later (the "GPL"), in
which case the provisions of the GPL are applicable instead of those above. If you wish to allow use of your version of this software only
under the terms of the GPL and not to allow others to use your version of this file under the MPL, indicate your decision by
deleting  the provisions above and replace  them with the notice and other provisions required by the GPL.  If you do not delete
the provisions above, a recipient may use your version of this software under either the MPL or the GPL.


Credits:
--------------------
eXtplorer is based on QuiXplorer 2.3.1 (available at http://quixplorer.sourceforge.net/).

eXtplorer makes use of the fabulous ExtJS Javascript Library by Jack Slocum, Ext JS, LCC (http://extjs.com/), and that does indeed explain
the name eXtplorer. 
Please note that ExtJS is released under the terms of the GNU General Public license version 3 (GPL v3). Because the MPL and
the GPL are not compliant to each other, you are not allowed to use the ExtJS library in eXtplorer if you decide to use and/or distribute eXtplorer
under the terms of the MPL.


----------------------------------------------------------------------------------------------------
Facts, you should know of:
* If you're running in trouble, because you don't have permissions to chmod() or
  write to files: That's a fact! Switch to FTP mode (or file mode if you're in ftp mode)
* eXtplorer is a multi-language Script.
  The Language is automatically picked from the browser (or from the global language in Joomla!/Mambo)
  
* eXtplorer comes with an User Management feature. As Joomla!/Mambo have their own framework, User Management isn't
	available in non-standalone mode.
* All users must login to access eXtplorer. You can allow visitors to view files by creating a "guest" account with "View" permissions.

* In Joomla!/Mambo the access to this  script is restricted to Super Administrators by default. If you want to change this,
  edit the file header of "admin.extplorer.php" and change the following:
  ****
		if (!$acl->acl_check( 'administration', 'config', 'users', $my->usertype )) {
			mosRedirect( 'index2.php', _NOT_AUTH );
		}
  ****
  to something different.
----------------------------------------------------------------------------------------------------

Troubleshooting:
	* Some browsers (e.g. Konqueror) may want to save a download as index.php.
	  To solve this, just supply the correct name when saving.
	* Internet Explorer may behave strangely when downloading files.
	  If you open the php-file download, the real download window should open.
	* Mozilla may add the extension 'php' to a file being downloaded.
	  Save as 'any file (*.*)' and remove the 'php' extension to get the proper name.
	  (NOTE: for php-files, this extension is correct)
	* If you are unable to perform certain operations,
	  try using an FTP-chmod to set the directories to 755 and the files to 644.
	* If you don't know the full name of a directory on your website,
	  you can use a php-script containing '<?php echo getcwd(); ?>' to get it.
	* The Search Function uses PCRE regex syntax to search; though wildcards like * and ?
	  should work (like with 'ls' on Linux), it may show unexpected behaviour.
	* User-management may logout unexpectedly or show other strange behaviour.
	  This is due to a bug in PHP 4.1.2; we would advise you to upgrade to a higher version.
----------------------------------------------------------------------------------------------------
Users:
	* You can easily magage users using the "admin" section of eXtplorer.
	* Standard, there is only one user: "admin", with password "admin";
	  you should change this password immediately.
----------------------------------------------------------------------------------------------------
Languages:
	* Users can select one of all available languages on login.
----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------
adapted for Joomla!: 	Soeren Eberhardt <soeren[aat]virtuemart.net>
original author:	the QuiX project
----------------------------------------------------------------------------------------------------


CHANGELOG.txt
****************************
Changelog for eXtplorer
Version $Id: CHANGELOG.txt 150 2009-06-27 20:10:10Z soeren $
****************************

- changed default state of "Show Directories" to enabled
- changed Edit Window to appear in a new Tab per File
- assigned Keyboard Events to the Grid (Delete, Ctrl-A, Ctrl-C, Ctrl-X)
- updated to ExtJS3 RC2
- updated EditArea to version 0.8.1.1 (loads faster and is more stable)

--- version 2.1.0 beta released ---

- allowing Download by users with "view" permission
- updated Services-JSON class to version 0.9.0
- updated GeSHi to version 1.0.8.3
- moved from ExtJS 1.1 to ExtJS 2.2

-- version 2.0.1 released (2009-01-15) ----

- added script for WebDAV access (disabled by default, requires 2 database tables and DB login credentials)
- fixed a security issue within script initialization

-- version 2.0.0 stable released (2008-08-05) ----

- added .csv to the editable file types
- when copying/moving multiple directories, only the first directory was processed
- fixed failing extraction of larger archives ("Failed to connect to the server")
- updated Editarea to 0.7.1.3
- fixed browsing & working on external FTP servers
- fixed dirselectors not working in FTP mode
- fixed file-mode switch link

-- version 2.0.0 RC4 released (2008-05-31) ----

- fixed installation package for Joomla! 1.5
- added Danish Language Files (thanks to Ronny Buelund!)

-- version 2.0.0 RC3 released (2008-05-31) ----

- [ 1944163 ] In germanf.php fehlt ein Eintrag
- fixed compatibility problem with JomComment+MyBlog (Services_JSON was redeclared)
- switched from "Codepress" to "EditArea" (http://www.cdolivet.net/index.php?page=editArea), which gives a lot of advantages:
	* faster loading of large files
	* built-in toggle feature
	* built-in Find, Search+Replace and Jump-To-Line Features
	
-- version 2.0.0 RC3 released (2008-05-26) ----

- fixed Cross-Site Scripting & File Disclosure Vulnerability
- fixed MimeType detection for search results

-- version 2.0.0 RC2 released (2008-02-10) ----

- fixed Standalone scripts.zip extraction error
- re-added System Info (moved to "About" Window)
- implemented basic UTF-8 support for file mode
- changed [ 1800028 ] Need to enlarge file edit box, or allow for full screen
- fixed [ 1791706 ] incorrect error reporting
- fixed [ 1790536 ] Browsing directory incorrect
- fixed [ 1782937 ] when unzipping a zip containing a zip: zip not present!

-- version 2.0.0 RC1 released ----

- fixed [ 1755938 ] status bar messages truncated when path is too long
- fixed [ 1759450 ] No textfield and browse button in upload file on IE7
- fixed [ 1762000 ] copy a file from a subfolder to "root" folder doesn't work
- fixed [ 1766233 ] Chmod error in FTP mode
- fixed [ 1761083 ] IE7 closing button & seach action (file search - subdirectories weren't included)

--- version 2.0.0 beta5 released ----

- added Swedish and Slovenian Language Files
- set "zip" as default archive type
- added message box that prevents eXtplorer usage on Joomla! 1.0.x versions >= 1.0.13 and explains why.
- added Joomla! version check
- implemented some changes for compatibility on Windows Systems with IIS running

--- version 2.0.0 beta4 released ----

- fixed header-only problems on Mambo/Joomla! < 1.0.10
- added Server-to-Server transfer capabilities (using fopen, cURL or fsockopen)
- fixed [#6092] Some strings remained hardcoded
- fixed [ 1754755 ] Save button when uploading file not translated
- fixed the Frontend Browsing part (when being used as a Joomla! component)
- updated finnish and french language file
- updated ExtJS to 1.1 RC1, included Konqueror Patch
- fixed [ 1752904 ] error on admin dialog opening
- fixed [ 1752901 ] Combo on Login page doesn't work in IE6 and 7
- fixed [ 1752534 ] Non-static method ext_Lang::msg()
- removed dialog_status from onHide function for the dialog
- fixed bugs in the language files  with undefined properties of non-existant var $_VERSION
- fixed bug with undefined var $acl

--- version 2.0.0 beta4 released ----

- fixed a bug which caused that a maximum of 50 directories in the same level were listed in the tree, not all
- fixed a bug with CodePress - didn't allow to edit files (when used in Joomla!)
- added a check to the standalone version to extract the contents of the file "scripts.zip" online and throw a warning if it doesn't succeed.
- removed the status bar from the Dialogs, moved the status bat into the Paging Toolbar
- added CSS styles to make the dropdown lists smaller

--- version 2.0.0 beta3 released ----

- re-activated User Management for the stand-alone version
- fixed a critical error which caused deletion of directories although "Cancel" was clicked
- fixed an error which prevented correct listing of numeric directories/files

--- version 2.0.0 beta2 released ----
- added double-click action which opens the Context Menu in the grid (Opera and Konqueror don't allow custom right-click menus)
- applied patches to ExtJS to allow usage with Konqueror
- added new ext_Lang class to be able to escape quotes and line endings for using it in JavaScript Strings
- fixed node context menu displaying outside of view
- added the fetchscript.php file to bundle and compress javascript and stylesheet files
  (Sending compressed js and css files significantly reduces the download size for ExtJS)

--- version 2.0.0 beta1 released ----

- fixed the "symlink points to target which can't be accessed" problem
- fixed "out of memory" problems
- directory names changed (no directories starting with a '.' anymore now)
- completely changed layout: directory tree, file grid with renaming, context menu, drag&drop
- added the ExtJS 1.0.1a library


--- version 1.6.0 released ----

- archives do not contain the whole file path anymore now, but only the relative one

include/edit.php
<?php
// ensure this file is being included by a parent file
if( !defined( '_JEXEC' ) && !defined( '_VALID_MOS' ) ) die( 'Restricted access' );
/**
 * @version $Id: edit.php 150 2009-06-27 20:10:10Z soeren $
 * @package eXtplorer
 * @copyright soeren 2007-2009
 * @author The eXtplorer project (http://sourceforge.net/projects/extplorer)
 * @author The	The QuiX project (http://quixplorer.sourceforge.net)
 * 
 * @license
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 * 
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 * 
 * Alternatively, the contents of this file may be used under the terms
 * of the GNU General Public License Version 2 or later (the "GPL"), in
 * which case the provisions of the GPL are applicable instead of
 * those above. If you wish to allow use of your version of this file only
 * under the terms of the GPL and not to allow others to use
 * your version of this file under the MPL, indicate your decision by
 * deleting  the provisions above and replace  them with the notice and
 * other provisions required by the GPL.  If you do not delete
 * the provisions above, a recipient may use your version of this file
 * under either the MPL or the GPL."
 * 
 * 
 */

/**
 * File-Edit Functions
 *
 */
class ext_Edit extends ext_Action {
	var	$lang_tbl = Array(
		'czech' => 'cs',
		'german' => 'de',
		'danish' => 'dk',
		'english' => 'en',
		'esperanto' => 'eo',
		'spanish' => 'es',
		'french' => 'fr',
		'croatian' => 'hr',
		'italian' => 'it',
		'japanese' => 'ja',
		'macedonian' => 'mk',
		'dutch' => 'nl',
		'polish' => 'pl',
		'portuguese' => 'pt',
		'russian' => 'ru',
		'slovenian' => 'sk'
	);

	function execAction($dir, $item) {		// edit file
		global $mainframe, $mosConfig_live_site;

		if(($GLOBALS["permissions"]&01)!=01) {
			ext_Result::sendResult('edit', false, ext_Lang::err('accessfunc' ));
		}
		$fname = get_abs_item($dir, $item);

		if(!get_is_file(utf8_decode($fname)))  {
			ext_Result::sendResult('edit', false, $item.": ".ext_Lang::err('fileexist' ));
		}
		if(!get_show_item($dir, $item)) {
			ext_Result::sendResult('edit', false, $item.": ".ext_Lang::err('accessfile' ));
		}

		if(isset($GLOBALS['__POST']["dosave"]) && $GLOBALS['__POST']["dosave"]=="yes") {
			// Save / Save As
			$item=basename(stripslashes($GLOBALS['__POST']["fname"]));
//			$item=stripslashes($GLOBALS['__POST']["fname"]);
			$fname2=get_abs_item($dir, $item);

			if(!isset($item) || $item=="") {
				ext_Result::sendResult('edit', false, ext_Lang::err('miscnoname' ));
			}
			if($fname!=$fname2 && @$GLOBALS['ext_File']->file_exists($fname2)) {
				ext_Result::sendResult('edit', false, $item.": ".ext_Lang::err('itemdoesexist' ));
			}

			$this->savefile($fname2);
			$fname=$fname2;

			ext_Result::sendResult('edit', true, ext_Lang::msg('savefile').': '.$item );

		}
		if(isset($GLOBALS['__POST']["doreopen"]) && $GLOBALS['__POST']["doreopen"]=="yes") {
			// File Reopen
			$extra = Array();
			$content = $GLOBALS['ext_File']->file_get_contents( $fname );
			if( get_magic_quotes_runtime()) {
				$content = stripslashes( $content );
			}

			$langs = $GLOBALS["language"];
			if ($langs == "japanese"){
				$_encoding = $GLOBALS['__POST']["file_encoding"];
				if ($content){
					$content = mb_convert_encoding($content, "UTF-8", $_encoding);
				}
				$extra["file_encoding"] = $_encoding;
			}

			$extra["content"] = $content;

			ext_Result::sendResult('edit', true, ext_Lang::msg('reopenfile').': '.$item, $extra);

		}

		// header
		$s_item=get_rel_item($dir,$item);	if(strlen($s_item)>50) $s_item="...".substr($s_item,-47);
		$id_hash = substr('f'.md5($s_item),0, 10);
		$s_info = pathinfo( $s_item );
		$s_extension = str_replace('.', '', $s_info['extension'] );
		switch (strtolower($s_extension)) {
			case 'txt':
				$cp_lang = 'text'; break;
			case 'cs':
				$cp_lang = 'csharp'; break;
			case 'css':
				$cp_lang = 'css'; break;
			case 'html':
			case 'htm':
			case 'xml':
			case 'xhtml':
				$cp_lang = 'html'; break;
			case 'java':
				$cp_lang = 'java'; break;
			case 'js':
				$cp_lang = 'javascript'; break;
			case 'pl': 
				$cp_lang = 'perl'; break;
			case 'ruby': 
				$cp_lang = 'ruby'; break;
			case 'sql':
				$cp_lang = 'sql'; break;
			case 'vb':
			case 'vbs':
				$cp_lang = 'vbscript'; break;
			case 'php':
				$cp_lang = 'php'; break;
			default: 
				$cp_lang = 'generic';
		}
	$content = $GLOBALS['ext_File']->file_get_contents( $fname );
	if( get_magic_quotes_runtime()) {
		$content = stripslashes( $content );
	}
	$cw = 250;
	$langs = $GLOBALS["language"];
	if ($langs == "japanese"){
		$cw = 200;
		if ($content){
			$_encoding = strtoupper(mb_detect_encoding($content, Array("ASCII", "ISO-2022-JP", "UTF-8", "EUCJP-WIN", "SJIS-WIN"), true));
			$content = mb_convert_encoding($content, "UTF-8", $_encoding);
			if ($_encoding == "SJIS-WIN"){
				$_encoding_label = "SJIS";
			} elseif ($_encoding == "EUCJP-WIN"){
				$_encoding_label = "EUC-JP";
			} elseif ($_encoding == "ISO-2022-JP"){
				$_encoding_label = "JIS";
			} elseif ($_encoding == "ASCII"){
				$_encoding_label = "UTF-8";
			} else {
				$_encoding_label = $_encoding;
			}
		} else {
			$_encoding_label = "UTF-8";
		}
	}
	?>
{
	"xtype": "form",
	"id": "<?php echo $id_hash ?>",
	"labelWidth": "300",
	"autoScroll": "true", 
	"url":"<?php echo basename( $GLOBALS['script_name']) ?>",
	"title": "<?php echo $GLOBALS["messages"]["actedit"].": $s_item" ?>",
	"frame": true,
	"closable": true,
	"items": [{

		"xtype": "textarea",
		"hideLabel": true,
		"name": "thecode",
		"id": "ext_codefield<?php echo $id_hash ?>",
		"fieldClass": "x-form-field",
		"value": '<?php echo str_replace(Array("\r", "\n"), Array('\r', '\n') , addslashes($content)) ?>',
		"width": "100%",
		"height": 300,
		"plugins": new Ext.ux.plugins.EditAreaEditor({
			"id" : "ext_codefield<?php echo $id_hash ?>",	
			"syntax": "<?php echo $cp_lang ?>",
			"start_highlight": true,
			"display": "later",
			"toolbar": "search, go_to_line, |, undo, redo, |, select_font,|, change_smooth_selection, highlight, reset_highlight, |, help" 
			<?php if (array_key_exists($langs, $this->lang_tbl)){?>
				,language: "<?php echo $this->lang_tbl[$langs] ?>"
				<?php 
				} ?>
		})
	},
	{
		
			width: <?php echo $cw ?>, 
			"xtype": "textfield",
			"fieldLabel": "<?php echo ext_Lang::msg('copyfile', true ) ?>",
			"name": "fname",
			"value": "<?php echo $item ?>",
			"clear": true
			}
<?php if ($langs == "japanese"){ ?>
			,{
			 "width": <?php echo $cw ?>,  
			 "style":"margin-left:10px", 
			 "clear":true,
			"xtype": "combo",
			"fieldLabel": "<?php echo ext_Lang::msg('fileencoding', true ) ?>",
			"name": "file_encoding",
			"store": [
						['UTF-8', 'UTF-8'],
						['SJIS-WIN', 'SJIS'],
						['EUCJP-WIN', 'EUC-JP'],
						['ISO-2022-JP','JIS']
					],
			"value" : "<?php echo $_encoding_label ?>",
			"typeAhead": true,
			"mode": "local",
			"triggerAction": "all",
			"editable": false,
			"forceSelection": true
			}
	
<?php } ?>
		],
	"buttons": [{
		"text": "<?php echo ext_Lang::msg('btnsave', true ) ?>", 
		"handler": function() {
			statusBarMessage( '<?php echo ext_Lang::msg('save_processing', true ) ?>', true );
			form = Ext.getCmp("<?php echo $id_hash ?>").getForm();
			form.submit({
				waitMsg: 'Saving the File, please wait...',
				//reset: true,
				reset: false,
				success: function(form, action) {
					datastore.reload();
					statusBarMessage( action.result.message, false, true );
				},
				failure: function(form, action) {
					statusBarMessage( action.result.error, false, false );
					Ext.Msg.alert('<?php echo ext_Lang::err('error', true) ?>!', action.result.error);
				},
				scope: form,
				// add some vars to the request, similar to hidden fields
				params: {option: 'com_extplorer', 
						action: 'edit', 
						code: editAreaLoader.getValue("ext_codefield<?php echo $id_hash ?>"),
						dir: '<?php echo stripslashes($dir) ?>', 
						item: '<?php echo stripslashes($item) ?>', 
						dosave: 'yes'
				}
			});
		}
	},{
		"text": "<?php echo ext_Lang::msg('btnreopen', true ) ?>", 
		"handler": function() { 
			statusBarMessage( '<?php echo ext_Lang::msg('reopen_processing', true ) ?>', true );
			form = Ext.getCmp("<?php echo $id_hash ?>").getForm();
			form.submit({
				waitMsg: 'Processing Data, please wait...',
				reset: false,
				success: function(form, action) {
					statusBarMessage( action.result.message, false, true );
					editAreaLoader.setValue("ext_codefield<?php echo $id_hash ?>", action.result.content);
				},
				failure: function(form, action) {
					statusBarMessage( action.result.error, false, false );
					Ext.Msg.alert('<?php echo ext_Lang::err('error', true) ?>!', action.result.error);
				},
				scope: form,
				// add some vars to the request, similar to hidden fields
				params: {
					option: 'com_extplorer', 
					action: 'edit', 
					dir: '<?php echo stripslashes($dir) ?>', 
					item: '<?php echo stripslashes($item) ?>', 
					doreopen: 'yes'
				}
			});
		}
	}]
}
	
<?php

	}
	function savefile($file_name) {			// save edited file
		if( get_magic_quotes_gpc() ) {
			$code = stripslashes($GLOBALS['__POST']["code"]);
		}
		else {
			$code = $GLOBALS['__POST']["code"];
		}
		$langs = $GLOBALS["language"];
		if ($langs == "japanese"){
			$_encoding = $GLOBALS['__POST']["file_encoding"];
			if ($_encoding != "UTF-8"){
				$code = mb_convert_encoding($code, $_encoding, "UTF-8");
			}
		}

		$res = $GLOBALS['ext_File']->file_put_contents( $file_name, $code );

		if( $res==false || PEAR::isError( $res )) {
			$err = basename($file_name).": ".ext_Lang::err('savefile' );
			if( PEAR::isError( $res ) ) {
				$err .= $res->getMessage();
			}
			ext_Result::sendResult( 'edit', false, $err );
		}

	}
}
//------------------------------------------------------------------------------
?>
